### PR TITLE
net: sockets: Fix datagram AF_PACKET socket handling

### LINKED ISSUE #45514: UDP Packet socket doesn't do L2 header processing
**Describe the bug**
Datagram socket (https://docs.zephyrproject.org/3.0.0/reference/kconfig/CONFIG_NET_SOCKETS_PACKET_DGRAM.html) should handle L2 header processing i.e., adding L2 header from sockaddr_ll in TX and removing L2 header and populating sockaddr_ll in the RX. But this is not being done and the packet is sent as is.

**To Reproduce**
Run the unit test: https://github.com/zephyrproject-rtos/zephyr/pull/45513 

**Expected behavior**
The unit test should pass.
**Impact**
Applications should do the extra L2 handling

**Logs and console output**

**Environment (please complete the following information):**
 - OS:Linux
 - Toolchain: Zephyr SDK
 - Commit SHA or Version used: 523cd60a77

### PR DESCRIPTION
**Pull request by rlubos**
This PR fixes the datagram AF_PACKET socket handling, which improperly received packet a packet before it was processed by L2 layer.

Additionally, fix the existing unit test to work correctly after update, so that the packets used in the test are not rejected by L2. Finally add two new unit tests for DGRAM sockets to improve test coverage.

Fixes #45514

### CODE REVIEW

**Comment by rlubos** [comment]
CC @krish2718

**Comment by krish2718** [comment]
LGTM, let me test and update.

**Review by jukkar — DISMISSED** [review]
Thanks Robert, LGTM

**Comment by krish2718** [comment]
I have ported this to ncs-zephyr (include/zephyr/net -> include/net) and tested with my app, the L2 packets are being dropped at `subsys/net/l2/ethernet/ethernet.c: ethernet_recv` because the type is not supported, my app is using a custom protocol for L2, so, I guess it needs to be patched as well?

**Comment by krish2718** [comment]
In [wpa_supplicant l2 packet ](https://github.com/nrfconnect/sdk-hostap/pull/1/files#diff-0f1819c18f80c5e9edc42583b253643fa0c393100d1414117d9ee1d8109933d2R153) an arbitrary protocol will be passed, in this specific case its EAPOL 0x888e.

**Comment by rlubos** [comment]
> In [wpa_supplicant l2 packet ](https://github.com/nrfconnect/sdk-hostap/pull/1/files#diff-0f1819c18f80c5e9edc42583b253643fa0c393100d1414117d9ee1d8109933d2R153) an arbitrary protocol will be passed, in this specific case its EAPOL 0x888e.

Ok, in such case we need to handle EAPoL in `ethernet_recv()` specifically, or maybe we could change the default behaviour in case DGRAM packet sockets are enabled, i. e., pass unrecognized EtherTypes as well to the application in such case. @jukkar Do you have an opinion on that?

**Comment by krish2718** [comment]
Yes, at Ethernet L2 level allow all packets, and then filter them at packet socket level based on protocol registered during socket open.

**Comment by carlescufi** [comment]
@krish2718 can you please approve if it fixes your issue?

**Comment by krish2718** [comment]
> @krish2718 can you please approve if it fixes your issue?

No, it doesn't fix my issue.

Before fix: packets received without L2 processsing
With fix: packets dropped during L2 processing, so, not received by app.

**Comment by rlubos** [comment]
> > @krish2718 can you please approve if it fixes your issue?
> 
> No, it doesn't fix my issue.
> 
> Before fix: packets received without L2 processsing 
> With fix: packets dropped during L2 processing, so, not received by app.

I think that a broken datagram socket processing and unsupported EtherType are two separate topics (especially that there were no mention of it in the original report), but I have added an extra commit here which allows to forward any Ethernet frame type to the application. Please give it a try.

**Comment by krish2718** [comment]
> > > @krish2718 can you please approve if it fixes your issue?
> > 
> > 
> > No, it doesn't fix my issue.
> > Before fix: packets received without L2 processsing
> > With fix: packets dropped during L2 processing, so, not received by app.
> 
> I think that a broken datagram socket processing and unsupported EtherType are two separate topics (especially that there were no mention of it in the original report), but I have added an extra commit here which allows to forward any Ethernet frame type to the application. Please give it a try.

The packets are coming now, but they are being treated as RAW packets (proto = ETH_P_ALL: through conn_raw_socket) and hence no L2 processing is done, L2 header is still intact. You can easily test this by modifying the testcase to set protocol as 0x888e.

**Comment by krish2718** [comment]
Apologies, I have picked up wrong ZEPHYR that didn't have your full changes, it works now.

**Comment by rlubos** [comment]
> Apologies, I have picked up wrong ZEPHYR that didn't have your full changes, it works now. Feel free to merge.

CC @carlescufi

File: subsys/net/l2/ethernet/ethernet.c [hunk 1]
```diff
@@ -254,8 +254,13 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
 		goto drop;
 #endif
 	default:
+#if defined(CONFIG_NET_ETHERNET_FORWARD_UNRECOGNISED_ETHERTYPE)
```
Comment (jukkar) [inline subsys/net/l2/ethernet/ethernet.c:257]: Minor nit, we could actually do `if (IS_ENABLED(....))` here.

Comment (rlubos) [reply subsys/net/l2/ethernet/ethernet.c]: Right, fixed.

**Review by jukkar — APPROVED** [review]

**Comment by krish2718** [comment]
I will close my PR https://github.com/zephyrproject-rtos/zephyr/pull/45513 as that test is covered here.

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**